apollo-router/src/plugins/authentication.rs

```diff
@@ -189,6 +204,136 @@ async fn get_jwks(url: Url) -> Option<JwkSet> {
+/// by the JWT header. The only mandatory parameter is "alg".
+/// Note: "none" is not implemented by jsonwebtoken, so it can't be part of the [`Algorithm`] enum.
+async fn search_jwks(
+    my_jwks: SharedDeduplicate,
+    criteria: &JWTCriteria,
+    jwks_urls: Vec<Url>,
+    context: &Context,
+) -> Result<Option<Jwk>, BoxError> {
+    const HIGHEST_SCORE: usize = 2;
+    let mut candidates = vec![];
+    for jwks_url in jwks_urls {
+        context.enter_active_request().await;
+        // Get the JWKS here
+        let jwks_opt = match my_jwks.get(jwks_url).await {
+            Ok(k) => k,
+            Err(e) => {
+                context.leave_active_request().await;
+                return Err(e.into());
+            }
+        };
+        context.leave_active_request().await;
+        let jwks = jwks_opt.ok_or(Error::JwkSet)?;
+        // Try to figure out if our jwks contains a candidate key (i.e.: a key which matches our
+        // criteria)
+        for mut key in jwks.keys {
+            let mut key_score = 0;
+            // We are only interested in keys which are used for signature verification
+            if let Some(purpose) = &key.common.public_key_use {
+                if purpose != &PublicKeyUse::Signature {
+                    continue;
+                }
+            } else if let Some(purpose) = &key.common.key_operations {
+                if !purpose.contains(&KeyOperations::Verify) {
+                    continue;
+                }
+            } else {
+                continue;
+            };
+
+            // Let's see if we have a specified kid and if they match
+            if criteria.kid.is_some() && key.common.key_id == criteria.kid {
+                key_score += 1;
+            }
+
+            // Furthermore, we would like our algorithms to match, or at least the kty
+            // If we have an algorithm that matches, boost the score
+            match key.common.algorithm {
+                Some(algorithm) => {
+                    if algorithm != criteria.alg {
+                        continue;
+                    }
+                    key_score += 1;
+                }
+                // If a key doesn't have an algorithm, then we match the "alg" specified in the
+                // search criteria against all of the algorithms that we support.  If the
+                // key.algorithm parameters match the type of parameters for the "family" of the
+                // criteria "alg", then we update the key to use the value of "alg" provided in
+                // the search criteria.
+                // If not, then this is not a usable key for this JWT
+                // Note: Matching algorithm parameters may seem unusual, but the appropriate
+                // algorithm details are not structured for easy consumption in jsonwebtoken and
+                // this is the simplest way to determine algorithm family.
+                None => match criteria.alg {
```

Comment:
that might work better as `match (criteria.alg, key.algorithm)`, especially with the elliptic curve case

Reply:
done: 53b6295c1934a75d51e24fcb8f198576af8f7fb4